api/api_ethereum.go

```diff
@@ -460,8 +460,13 @@ func (api *EthereumAPI) GetHeaderByHash(ctx context.Context, hash common.Hash) m
 // * When fullTx is true all transactions in the block are returned, otherwise
 //   only the transaction hash is returned.
 func (api *EthereumAPI) GetBlockByNumber(ctx context.Context, number rpc.BlockNumber, fullTx bool) (map[string]interface{}, error) {
+	// Klaytn backend returns error when there is no matched block but
+	// Ethereum returns it as nil without error, so we should return is as nil when there is no matched block.
 	klaytnBlock, err := api.publicBlockChainAPI.b.BlockByNumber(ctx, number)
-	if klaytnBlock != nil && err == nil {
+	if err != nil {
```

Comment:
@aeharvlee What if other error occurs? There is no way to return err?!

Reply:
@kjhman21 
Currently only one error which is `fmt.Errorf("the block does not exist (block number: %d)", blockNr)` can be returned from [CNAPIBackend.BlockByNumber](https://github.com/klaytn/klaytn/blob/8916320ea83d521e35d64518d90184406f4080b0/node/cn/api_backend.go#L123-L138). 

Ethereum treats that case(= The block does not exist) as `nil`. 

By the way, there is [the case](https://github.com/klaytn/klaytn/blob/8916320ea83d521e35d64518d90184406f4080b0/api/api_ethereum.go#L465) when error can occurs during `EthereumAPI.GetBlockByNumber` and `EthereumAPI.GetBlockByHash`.
Both APIs use [rpcMarshalBlock](https://github.com/klaytn/klaytn/blob/8916320ea83d521e35d64518d90184406f4080b0/api/api_ethereum.go#L1457-L1488) to marshal Block data structure to `map[string]interface{}`.
[rpcMarshalBlock](https://github.com/klaytn/klaytn/blob/8916320ea83d521e35d64518d90184406f4080b0/api/api_ethereum.go#L1457-L1488) calls [rpcMarshalHeader](https://github.com/klaytn/klaytn/blob/8916320ea83d521e35d64518d90184406f4080b0/api/api_ethereum.go#L1413-L1455) internally and [rpcMarshalHeader](https://github.com/klaytn/klaytn/blob/8916320ea83d521e35d64518d90184406f4080b0/api/api_ethereum.go#L1413-L1455) try to fetch block proposer using [backend.Author](https://github.com/klaytn/klaytn/blob/8916320ea83d521e35d64518d90184406f4080b0/consensus/istanbul/backend/engine.go#L118-L121) at `istanbul/backend/engine.go`. 
If error occurs during [backend.Author](https://github.com/klaytn/klaytn/blob/8916320ea83d521e35d64518d90184406f4080b0/consensus/istanbul/backend/engine.go#L118-L121), it returns error so `EthereumAPI.GetBlockByNumber` and `EthereumAPI.GetBlockByHash` returns error in that case. This is the case when other error can occur during those APIs and this it the different part from Ethereum.

cc. @aidan-kwon

Reply:
Either one is okay for now. But, I think error string comparison increases the complexity of these APIs.